modules/drivers/audio/speaker_component.cc

```diff
@@ -0,0 +1,112 @@
+/******************************************************************************
+ * Copyright 2017 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/drivers/audio/speaker_component.h"
+
+namespace apollo {
+namespace drivers {
+namespace audio {
+
+bool SpeakerComponent::Init() {
+  speaker_config_ptr_ = std::make_shared<SpeakerConfig>();
+  if (!apollo::cyber::common::GetProtoFromFile(config_file_path_,
+                                               speaker_config_ptr_.get())) {
+    return false;
+  }
+  AINFO << "Speaker config: " << speaker_config_ptr_->DebugString();
+
+  // new speaker device
+  speaker_device_ptr_.reset(new Respeaker());
+  speaker_device_ptr_->init(speaker_config_ptr_);
+
+  // dump config, calculate size and reserve buffer for audio
+  // chunk_: number of frames per chunk; chunk_size_: number of bytes per chunk
+  n_channels_ = speaker_config_ptr_->channel_type_size();
+  chunk_ = speaker_config_ptr_->chunk();
+  total_frames_ = speaker_config_ptr_->sample_rate() *
+                  speaker_config_ptr_->record_seconds();
+  chunk_size_ = speaker_device_ptr_->stream_ptr_->get_chunk_size(chunk_);
+
+  buffer_size_ = speaker_device_ptr_->stream_ptr_->get_chunk_size(total_frames_);
+  buffer_ = reinterpret_cast<char *>(malloc(buffer_size_));
```

Comment:
Better use c++ style `new[]`  instead of c-style malloc for memery allocation.

Reply:
Yes, didn't notice that. Have changed it to `new`